[FSDP] Do not clone buffers; offload buffers to CPU if needed

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#77000 [FSDP] Do not clone buffers; offload buffers to CPU if needed**
* #76994 [FSDP] Move param/buffer name comp. to ctor for `ignored_modules`

This avoids cloning buffers for full model state dict and makes sure to offload buffers to CPU if the state dict config has `offload_to_cpu=True`.

Sadly, this PR undoes some work from the PR underneath in the stack, but I split the two for clarity.

Differential Revision: [D36271736](https://our.internmc.facebook.com/intern/diff/D36271736)

Commit:
[FSDP] Do not clone buffers; offload buffers to CPU if needed

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -478,7 +493,10 @@ def test_state_dict_with_ignored_modules(self):
-        for tensor, tensor_name in ignored_tensor_to_tensor_name.items():
+        for tensor, tensor_name in {
+            **ignored_tensor_to_tensor_name,
+            **buffer_to_buffer_name,
```
I'm assuming that previously including buffers in this check would have raised an error?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1544,10 +1537,20 @@ def _full_post_state_dict_hook(
+            # Do not need to clone buffers
+            if clean_key in self._buffer_names:
+                # Offload the buffer to CPU if needed since it is not done in
+                # `_summon_full_params()`
```
should we eventually push this down to `summon_full_params` so that we get it for free? Feel free to file an issue if you feel that would be better than doing it manually here. It may also make things more consistent with `summon_full_params`.

awgu replied:
It would be clean if we could do this in `summon_full_params()`, but in my understanding, there is one main barring factor:
If we do it in `summon_full_params()`, then we need to destructively modify the actual buffer variable's data to point to CPU storage before the `state_dict()` logic triggers and restore the data to point back to GPU storage after the `state_dict` has been computed. Without some care, the original GPU storage will get freed and later re-allocated, which is wasteful.

I chose to do it in the full post state dict hook since then we only incur the necessary cost of allocating CPU storage for the buffer's data. The original buffer's data pointing to GPU storage is not freed and hence does not need to be reallocated.

rohan-varma reviewed (approved):
Looks good, thanks! In the future, could we try to have a single PR for each change just so that they are easier to understand? i.e.  a PR for not cloning buffer and a separate one for offloading them.

Also, how do we unittest that buffers are offloaded to CPU? I don't appear to see any unittest changes for that? Stamping to unblock

Commit:
Update on "[FSDP] Do not clone buffers; offload buffers to CPU if needed"


This avoids cloning buffers for full model state dict and makes sure to offload buffers to CPU if the state dict config has `offload_to_cpu=True`.

Sadly, this PR undoes some work from the PR underneath in the stack, but I split the two for clarity.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36271736).

awgu commented:
Failure is due to `test_variant_consistency_eager_eig_cuda_complex64`, which is unrelated.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1423,11 +1419,12 @@ def _setup_streams(self) -> None:
-                # buffer names since only the root's names are fully prefixed
-                # like the state dict keys
+                # execution order data structure and the root's ignored
+                # parameters and all buffer names since only the root's names
```
```suggestion
                # parameter names and all buffer names since only the root's names
```
I need to fix this grammatical typo I introduced last night, but I am not going to re-push and re-trigger CI just for this. (Leaving this comment as a reminder to do so.)

awgu commented:
@pytorchbot merge this please
